libbeat/processors/add_nomad_metadata/nomad.go

```diff
@@ -0,0 +1,166 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package add_nomad_metadata
+
+import (
+	"time"
+
+	nomad "github.com/hashicorp/nomad/api"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	nomadlib "github.com/elastic/beats/libbeat/common/nomad"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+	"github.com/elastic/beats/libbeat/processors/actions"
+	jsprocessor "github.com/elastic/beats/libbeat/processors/script/javascript/module/processor"
+)
+
+const (
+	processorName        = "add_nomad_metadata"
+	nomadRegexDataPrefix = "nomad."
+	timeout              = time.Second * 5
+)
+
+type nomadAnnotator struct {
+	log *logp.Logger
+
+	sourceProcessor processors.Processor
+	cache           *cache
+	datacenter      string
+	metaPrefix      string
+}
+
+func init() {
+	processors.RegisterPlugin(processorName, New)
+	jsprocessor.RegisterPlugin("AddNomadMetadata", New)
+}
+
+// New constructs a new add_kubernetes_metadata processor.
+func New(cfg *common.Config) (processors.Processor, error) {
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrapf(err, "fail to unpack the %v configuration", processorName)
+	}
+
+	client, err := nomadlib.NewClient(config.Address, config.Region, config.SecretID, nil)
+	if err != nil {
+		logp.Err("nomad: Couldn't create nomad client: %v", err)
+		return nil, err
+	}
+
+	nodeID := config.Node
+	if nodeID == "" {
+		if nodeID, err = nomadlib.GetLocalNodeID(client); err != nil {
+			logp.Err("nomad: Couldn't get nomad node ID: %v", err)
+			return nil, err
+		}
+	}
+	node, _, err := client.Nodes().Info(nodeID, nil)
+	if err != nil {
+		logp.Err("nomad: Couldn't get node information: %v", err)
+		return nil, err
+	}
+
+	var procConf, _ = common.NewConfigFrom(map[string]interface{}{
+		"field":  "log.file.path",
+		"prefix": nomadRegexDataPrefix,
+		"regexp": `.*/(?P<allocation>.+)/alloc/logs/(?P<task>.+)\.(?P<stream>std.+)\.[0-9]+`,
+	})
+	sourceProcessor, err := actions.NewExtractRegexp(procConf)
+	if err != nil {
+		return nil, err
+	}
+
+	annotator := &nomadAnnotator{
+		log:             logp.NewLogger(processorName),
+		sourceProcessor: sourceProcessor,
+		cache:           newCache(config.CleanupTimeout),
+		datacenter:      node.Datacenter,
+		metaPrefix:      config.MetaPrefix,
+	}
+
+	watcher, err := nomadlib.NewWatcherWithClient(nomadlib.WrapClient(client), nodeID, annotator.modifiedAllocation)
+	if err != nil {
+		logp.Err("nomad: Couldn't create allocation watcher")
+		return nil, err
+	}
+
+	go watcher.Start()
```

Comment:
Which function calls `watcher.Stop`? How do you make sure that the goroutine is stopped when it has to?

Reply:
It's not called anywhere. I tried to find a way to call this, but the [Processor](https://github.com/elastic/beats/blob/master/libbeat/processors/processor.go#L38-L41) interface only defines the _Run_ and _String_ methods. Anyway, in filebeat, the processor is defined on the configuration and applied until the program is stopped, so `watcher.Stop` would only be called at the end of the program, and in that case the program will exit, so no need to clean goroutines. 

I was looking at other processors (docker, k8s) and the Stop functions of them are not called on the provider.

What should I do here?